src/transform/expressionEvaluation.ml

```diff
@@ -0,0 +1,199 @@
+        (* Pretty print *)
+        |> Cil.d_stmt ()
+        |> Pretty.sprint ~width:0
+        (* Split into lines *)
+        |> String.split_on_char '\n'
+        (* Remove preprocessor directives *)
+        |> List.filter (fun line -> line.[0] <> '#')
+        (* Remove indentation *)
+        |> List.map String.trim
+        (* Concatenate lines into one *)
+        |> List.fold_left (^) ""
+
+    class evaluator (file : Cil.file) (ask : Cil.location -> Queries.t -> Queries.Result.t) =
+      object (self)
+
+        val global_variables =
+          file.globals
+            |> List.filter_map (function Cil.GVar (v, _, _) -> Some (v.vname, Cil.Fv v) | _ -> None)
+        val statements =
+          file.globals
+            |> List.filter_map (function Cil.GFun (f, _) -> Some f | _ -> None)
+            (* Take all statements *)
+            |> List.map (fun (f : Cil.fundec) -> f.sallstmts |> List.map (fun s -> f, s))
+            |> List.flatten
+            (* Add locations *)
+            |> List.map (fun (f, (s : Cil.stmt)) -> (Cil.get_stmtLoc s.skind, f, s))
+            (* Filter artificial ones by impossible location *)
+            |> List.filter (fun ((l : Cil.location), _, _) -> l.line >= 0)
+            (* Create hash table *)
+            |> List.fold_left (fun ss (l, f, s) -> Hashtbl.add ss l (f, s); ss) (Hashtbl.create 0)
+
+        method evaluate location expression_string =
+          (* Compute the available local variables *)
+          let local_variables =
+            match Hashtbl.find_option statements location with
+            | Some (function_definition, _) -> function_definition.slocals |> List.map (fun (v : Cil.varinfo) -> v.vname, Cil.Fv v)
+            | None -> []
+          in
+          (* Parse expression *)
+          match ~? (fun () -> Formatcil.cExp expression_string (local_variables @ global_variables)) with
+            (* Expression unparseable at this location *)
+          | None ->
+              if is_debug () then print_endline "| (Unparseable)";
+              Some false
+            (* Successfully parsed expression *)
+          | Some expression ->
+              (* Evaluate at (directly before) the location *)
+              match self#try_ask location expression with
+                (* Dead code or not listed as part of the control flow *)
+              | None ->
+                  if is_debug () then print_endline "| (Unreachable)";
+                  Some false
+                (* Valid location *)
+              | Some value_before ->
+                  (* Use the last listed matching statement; TODO: use Hashtbl.find_all to consider other matching statements *)
+                  let _, statement = Hashtbl.find statements location in
+                  (* Use the first evaluable successor; TODO: consider other successors *)
+                  let succeeding_statement = ref None in
+                  let successor_evaluation =
+                    List.find_map_opt
```

Comment:
This seems to be a recent addition `OCaml >= 4.10.0`, but we aim to remain compatible with `4.07.1` (but only added checks for this to the CI in 46602ca1b696b138c6a7ae2030cd7749ecd0c453, this is why this PR still passes).

Can you make this compatible with `4.07.1`?

Reply:
The `4.07.1` checks seem to pass somehow, but I can't explain why.